Reduce access to invalid communicator as much as possible.

kronbichler opened the pull request:
Update:
Fixes #3656.

I'm now preventing the access of the invalid states in the test and additionally fix the underlying problem that we do not duplicate communicators any more so it is the user's responsibility to keep the communicator alive and we do not have to check (in C++ or MPI) that a given MPI communicator object was indeed deleted and our Trilinos vector needs to do something clever in `operator=`. I invite the people who could reproduce the problem in #3656, i.e., @drwells, @bangerth, @masterleinad, to have a look at this PR whether it fixes the problem.

I will leave a note in #745 that we should not use any old state in `operator=` and `reinit()` for the new vector interfaces because I do not think we can avoid a problem in our interfaces in case one really wants to construct a case for it.

masterleinad commented:
I can confirm that running with `valgrind` the old code aborted while the new one does not. Unfortunately, I seem to fail to reproduce the failure of the original test case without calling `valgrind`.
Nevertheless, this seems to solve this particular issue.

kronbichler commented:
As a measure of safety, according to #3656 I would also suggest to delete all growing vector memory instances when calling MPI_Comm_free in `distributed/shared_tria.cc`.

bangerth commented on source/lac/trilinos_vector.cc:
```diff
@@ -504,6 +505,24 @@ namespace TrilinosWrappers
+      // Need to ask MPI whether the communicators are the same. We would like
+      // to use these checks but currently we cannot make sure the memory of
+      // my_comm is not stale from some MPI_Comm_free somewhere. Thus, the
+      // following code is commented out.
```
can you specifically reference `GrowingVectorMemory` to make clear what use case you have in mind?

kronbichler replied:
I will do that. Should we go along an also implement a function
```
namespace internal
{
  template <typename VectorType>
  void release_unused_growing_vector_memory()
  {
    GrowingVectorMemory<VectorType> mem;
    mem.release_unused_memory();
  }
}
```
that we instantiate for all of our vector types and call at the two places where we have `MPI_Comm_free` (alongside with a comment on why we do that).

kronbichler replied:
Wait, I of course did not mean to make this a template on the release function but rather let our template instantiation machinery list all `GrowingVectorMemory<VectorType>` instances in the `vector_memory.cc` file.

kronbichler commented:
@bangerth I now added the fix discussed in #3656 that deletes all unused vectors. This should now be a safe solution.

bangerth commented:
You'll have to fix the indentation.

bangerth commented on source/distributed/tria_base.cc:
```diff
@@ -72,6 +73,10 @@ namespace parallel
         other_tria_x = dynamic_cast<const dealii::parallel::Triangulation<dim,spacedim> *>(&other_tria))
       {
+        // make sure that unusued vectors possibly using this communicator do
+        // not end up with a wrong communicator
```
wrong -> stale

bangerth commented on source/distributed/tria_base.cc:
```diff
@@ -72,6 +73,10 @@ namespace parallel
     if (const dealii::parallel::Triangulation<dim,spacedim> *
         other_tria_x = dynamic_cast<const dealii::parallel::Triangulation<dim,spacedim> *>(&other_tria))
       {
+        // make sure that unusued vectors possibly using this communicator do
```
-> unused

bangerth commented on source/distributed/tria_base.cc:
```diff
@@ -99,6 +104,10 @@ namespace parallel
   {
 #ifdef DEAL_II_WITH_MPI
+    // make sure that unusued vectors possibly using this communicator do
+    // not end up with a wrong communicator
```
same fixes here

bangerth commented on source/lac/vector_memory_release.inst.in:
```diff
@@ -0,0 +1,34 @@
+{
+    \{
+    dealii::GrowingVectorMemory<dealii::VECTOR> memory;
+    memory.release_unused_memory();
```
This is a static function -- you can just call it on the class, no need to create an object.

kronbichler commented:
@bangerth I now remove the duplication from the tria class. Can you have a look at it? If you agree, I will squash some commits tomorrow and add a note in changes.h.

bangerth commented on source/distributed/tria_base.cc:
```diff
@@ -43,10 +44,9 @@ namespace parallel
+    mpi_communicator (mpi_communicator),
     my_subdomain (Utilities::MPI::this_mpi_process (this->mpi_communicator)),
-    n_subdomains(Utilities::MPI::n_mpi_processes(mpi_communicator))
+    n_subdomains(Utilities::MPI::n_mpi_processes (this->mpi_communicator))
```
I was about to say "Ohh, good catch!" until I realized that it makes absolutely no difference :-)

kronbichler replied:
I just put it there for consistency... ;-)

bangerth commented:
Yes, this looks good. You took the call to `release_all_memory()` out again, so in principle the function is no longer needed. That said, I think it was not unreasonable to release memory when a triangulation is refined -- the size of all vectors is going to change anyway, so why hold on to vectors we can no longer use as is. Want to put the call back in? Or connect the function to the `pre_refinement` signal of each triangulation object (in the constructor of `Triangulation`)? Not sure I know which way to go, but it's worth a discussion.

kronbichler commented:
I now put the release of memory back into the triangulation. I also cleaned up my commits and added a test to ensure that the cleanup really is effective.
